### ISSUE TITLE
Using LBFGS or Levenberg-Marquardt to train neural network

### ISSUE THREAD

**Issue by e-eight** [davisking/dlib#2087]
I am trying to train a two-layer feedforward neural network with a very small amount of data. I have seen that in such cases using a second order optimization method like Levenberg-Marquardt or LBFGS gives better results than SGD or Adam. On Matlab I can easily do it using `trainlm` or `trainbr`. Even though I see that `dlib` has implementations for LBFGS and Levenberg-Marquardt as optimizers, for training of neural networks only SGD and Adam are available. I was wondering if there is any way I can use say the `solve_least_squares_lm` method to train the neural network.

**Comment by davisking** [comment]
LBFGS will usually badly overfit.  There is some not well understood regularization property of SGD like methods that is very beneficial.  I find this particularly frustrating since LBFGS and methods like it are far superior to SGD in terms of minima finding, but it is what it is.

But you could certainly do this anyway, you just need to be able to read the parameters of a network and then assign them back.  With that you can plug into an optimizer like LBFGS.  How you get parameters of a network and work with network objects is all documented, so see the docs.  Start with http://dlib.net/dnn_introduction_ex.cpp.html, but also don't neglect the reference docs for the API, which document every function and class.

**Comment by davisking** [comment]
I should also note that if you are just training a simple 2 layer network you should instead use a kernel method.  Kernel methods also train 2 layer networks, but they do so in a theoretically sound and very numerically robust and fast way.  So using KRR or SVM or whatever you need with a reasonable kernel like the RBF kernel is the way to go.

**Comment by e-eight** [comment]
> LBFGS will usually badly overfit. There is some not well understood regularization property of SGD like methods that is very beneficial. I find this particularly frustrating since LBFGS and methods like it are far superior to SGD in terms of minima finding, but it is what it is.
> 
> But you could certainly do this anyway, you just need to be able to read the parameters of a network and then assign them back. With that you can plug into an optimizer like LBFGS. How you get parameters of a network and work with network objects is all documented, so see the docs. Start with http://dlib.net/dnn_introduction_ex.cpp.html, but also don't neglect the reference docs for the API, which document every function and class.

Thanks, I will check that out. I am actually more interested in using Levenberg-Marquardt than LBFGS, since I am minimizing the mean squared error, and on MATLAB I see that Levenberg-Marquardt gives better results.



> I should also note that if you are just training a simple 2 layer network you should instead use a kernel method. Kernel methods also train 2 layer networks, but they do so in a theoretically sound and very numerically robust and fast way. So using KRR or SVM or whatever you need with a reasonable kernel like the RBF kernel is the way to go.

Thanks for the suggestion. I am indeed investigating what other methods than a 2 layer network can I use.

**Comment by davisking** [comment]
To be clear, a kernel SVM and KRR *are* two lay neural networks.  They are special solvers that optimize them very rapidly and reliably.  Levenberg-Marquardt has nothing on the speed and accuracy of the kernel method solvers.  Like really, it is a mistake to use Levenberg-Marquardt to solve this kind of problem, unless it's the case that you have a special flavor of 2 layer NN that isn't amenable to kernel methods.  You haven't said exactly what you are doing, but for instance, if you are feeding a small number of variables as input to the NN then a gaussian kernel network would probably be a good fit, and those are readily handled by kernel methods.

**Comment by e-eight** [comment]
I am trying to replicate the work done in [this paper](https://arxiv.org/pdf/1810.04009.pdf) (Figure 1 and Figure 2 particularly). This is some research done by one of the previous members of our group, but the implementation is in MATLAB, and there are some serious performance issues. Not sure if that is due to MATLAB, or because this is not the ideal approach to this problem.

We need a method that generalizes well on small data. So what was done in that work was a large ensemble of shallow 2-layer networks were trained on the data, each network was trained on a random sub selection of the data. Only a fraction of the trained networks were retained for prediction, based on their test mean squared error. For the kind of problems we are looking at, this method works, but there's the performance issue.

I am looking at gaussian processes, and kernel ridge regression but not sure which one will give similar results. The reason I mentioned Levenberg-Marquardt was that MATLAB seems to use that internally for training such networks.
